docs/modal.rst

```diff
@@ -0,0 +1,69 @@
+Modals
+============
+
+You can use modal forms as an alternative form to gather user input.
+To do this, you have to return a :class:`.Modal` object instead of a string or a Message as the result of an interaction callback.
+
+Constructing a Modal form
+---------------------------------
+
+In order to show a Modal to an user, you must return either an instance of Modal class or subclass it yourself.
+To construct them, you have to pass a Custom ID, the modal's Title
+and the fields, which must be a list of :class:`.ActionRow`s.
+
+.. code-block:: python
+
+    @discord.command()
+      def modal_example(ctx):
+          fields = [
+            ActionRow(
+                TextInput(
+                    "user_input_name",
+                    "Enter your name"
+                )
+            )
+          ]
+          return Modal("example_modal", "Hello there", fields)
+
+
+Defining a modal handler
+--------------------------------
+
+Modal handlers are callbacks that receive class:`.Context` objects and
+can return messages as usual. However, you cannot return another Modal
+as the response for a Modal.
+
+They are defined the same way you would register a callback for a Button or another component:
+
+.. code-block:: python
+
+    @discord.command()
+      def modal_example(ctx):
+          fields = [
+            ActionRow(
+                [
+                    TextInput(
+                        "user_input_name",
+                        "Enter your name"
+                    )
+                ]
+            )
+          ]
+          return Modal("example_modal", "Hello there", fields)
+
+
+    @discord.custom_handler("example_modal")
+    def modal_callback(ctx):
+        components = ctx.components
```

Comment:
This is pretty verbose, I can see now why you wanted to flatten the `Context.components` to remove the action rows. I do think we can still make this cleaner without resorting to that though. Maybe a method like `Context.get_component("...")` which would take in a custom ID string and return the first component in the tree that matches it?

Reply:
Looks good!